Feature/process folder

ohadravid opened the pull request:
For CR

omerbenamram committed:
started work on postprocessor

omerbenamram committed:
acoustid tagger working
simple tagger working
added test

omerbenamram committed:
harder, better, faster, stronger AcoustID tagger

omerbenamram committed:
modified test to match changes

omerbenamram committed:
dropped usage of beets internal utils entierly

omerbenamram committed:
modified simpletagger

omerbenamram committed:
there isn't much left of the original logic...

omerbenamram committed:
more work on post processor structure

omerbenamram committed:
changed name of folder

omerbenamram committed:
some more refactoring

omerbenamram committed:
renamer now working

omerbenamram committed:
using py.path istead of pathlib python 2.7 port since it caused problems with unicode

omerbenamram committed:
simplified design by removing global processors list

omerbenamram committed:
cleaned leftoves

omerbenamram committed:
small docu

ohadravid commented on src/headphones2/postprocess/component_base.py:
```diff
@@ -0,0 +1,13 @@
+class PostProcessorComponentBase(object):
+    __metaclass__ = abc.ABCMeta
+
+    @staticmethod
```
Using static method in this context is not recommended.
You will lose the ability to have a state for your processor, and it makes for a strange api.

ohadravid commented on src/headphones2/postprocess/component_base.py:
```diff
@@ -0,0 +1,13 @@
+
+    @staticmethod
+    def process(item_list, **kwargs):
+        raise NotImplementedError
```
Always use the () form of exception raising.

omerbenamram replied:
fixed :)

ohadravid commented on src/headphones2/postprocess/extensions/renamer.py:
```diff
@@ -0,0 +1,72 @@
+            original_path = Path(item.path)
+            new_name = name_string
+
+            for param, value in components.iteritems():
```
There must be a better way! Why not use 'format()' and named places ('{artist} {album}...')

omerbenamram replied:
from string import Template
bam.

ohadravid commented on src/headphones2/postprocess/process_folder.py:
```diff
@@ -0,0 +1,71 @@
+from headphones2.postprocess.component_base import PostProcessorException
+from headphones2.postprocess.extensions.renamer import Renamer
+from headphones2.postprocess.taggers.acoustid_tagger import AcoustIDAlbumTagger
+from headphones2.postprocess.taggers.simple_tagger import SimpleBeetsTagger
```
Fix imports to be less verbose using __init__

omerbenamram replied:
done

ohadravid commented on src/headphones2/postprocess/process_folder.py:
```diff
@@ -0,0 +1,71 @@
+        last = val
+    yield last, True
+
+class FolderIterator(object):
```
Why not use globe?

omerbenamram replied:
so - glob gave me lots of trouble due to its interal calling to fnmatch and re (very strage behavior) with directories that have [] chars in them for exmaple.

went with a simple filter approach:
    for ext in extensions:
        matches = p.listdir(fil=lambda x: x.ext == ext)
        file_pathes_to_process.extend(map(str, matches))

removed the folder class

ohadravid commented on src/headphones2/postprocess/process_folder.py:
```diff
@@ -0,0 +1,71 @@
+    try:
+        beets_tagger.process(list_of_items, expected_artist=expected_artist, expected_album=expected_album)
+    except PostProcessorException:  # simple tagging flow failed
+        logger.debug('Calling AcoustID Tagger')
```
Use logger.exception (which will save the stacktrace)

ohadravid commented on src/headphones2/postprocess/process_folder.py:
```diff
@@ -0,0 +1,71 @@
+        if is_success:
+            beets_tagger.process(list_of_items, expected_artist=expected_artist, expected_album=expected_album, expected_release_id=recommendation)
+            return True
+        raise  # both our taggers failed :(
```
Raise a new, custom exception

omerbenamram replied:
:)

ohadravid commented on src/headphones2/postprocess/process_folder.py:
```diff
@@ -0,0 +1,71 @@
+
+def _tag_album_and_fix_metadata(list_of_items, expected_artist=None, expected_album=None):
+    aid_tagger = AcoustIDAlbumTagger()
+    beets_tagger = SimpleBeetsTagger()
```
Pass beets_tagger to AcoustIDAlbumTagger tagger, so that the 'except' flow is handled by it entirely. (It try to send the sample, if ok, use the beets tagger).

ohadravid commented on src/headphones2/postprocess/process_folder.py:
```diff
@@ -0,0 +1,71 @@
+            return True
+        raise  # both our taggers failed :(
+
+    return True
```
Why return true? If we fail we throw and exception. If you wan silent exceptions, return false and do not raise.

omerbenamram replied:
fixed

ohadravid commented on src/headphones2/postprocess/taggers/acoustid_tagger.py:
```diff
@@ -0,0 +1,117 @@
+        determines which releases the items have in common.
+        also uses number of tracks to make sure its the same release
+        """
+        result_ids = chain.from_iterable([result.release_id for result in result_list if result])
```
Why chain? Counter works for simple lists as well.

ohadravid commented on src/headphones2/postprocess/taggers/acoustid_tagger.py:
```diff
@@ -0,0 +1,117 @@
+        for release_id, _ in top_matches:
+            r = musicbrainzngs.get_release_by_id(release_id, includes='recordings')
+            mediums = r['release']['medium-list']
+            tracks = sum([mediums[i]['track-count'] for i, j in enumerate(mediums)])
```
Why not `sum([medium['track-count'] for medium in mediums])`?

ohadravid commented on src/headphones2/postprocess/taggers/acoustid_tagger.py:
```diff
@@ -0,0 +1,117 @@
+        """
+        result_ids = chain.from_iterable([result.release_id for result in result_list if result])
+        top_matches = Counter(result_ids).most_common(MAX_RELEASES)
+        num_of_tracks = len(result_list)
```
A better name would be `actual_number_of_tracks` of 'expected_number_of_tacks`.

ohadravid commented on src/headphones2/postprocess/taggers/acoustid_tagger.py:
```diff
@@ -0,0 +1,117 @@
+
+    Result = namedtuple('Result', ['fingerprint', 'acoustid', 'recording_id', 'release_id'])
+
+    @staticmethod
```
Change everything to be instance based and not static.

ohadravid commented on src/headphones2/postprocess/taggers/acoustid_tagger.py:
```diff
@@ -0,0 +1,117 @@
+        try:
+            res = acoustid.lookup(API_KEY, fingerprint, duration, meta='recordings releases')
+        except acoustid.AcoustidError as exc:
+            logger.debug(u'fingerprint matching {0} failed: {1}', filepath, exc)
```
logger.exception will save the stacktrace to the log.

ohadravid commented on src/headphones2/postprocess/taggers/acoustid_tagger.py:
```diff
@@ -0,0 +1,117 @@
+            logger.debug(u'fingerprint matching {0} failed: {1}', filepath, exc)
+            return None
+
+        logger.debug(u'fingerprinted {0}', filepath)
```
All those 'u's are strange. Just add 'unicode_literals' from future?

omerbenamram replied:
this was actually original code from beets. changed it to be plain unicode

ohadravid commented on src/headphones2/postprocess/taggers/acoustid_tagger.py:
```diff
@@ -0,0 +1,117 @@
+
+        logger.debug(u'matched recordings {0} on releases {1}', recording_ids, release_ids)
+
+        return AcoustIDAlbumTagger.Result(fingerprint, acoustid_result['id'], recording_ids, release_ids)
```
I would place `Result` in the module level, since it is more common and is an external api.
I will also decuple this function from the name of the class, without using 'self.Result' or 'cls.Result' which is even less common.

omerbenamram replied:
moved it to the module leve. are any other changes necessary?

ohadravid commented on src/headphones2/postprocess/taggers/acoustid_tagger.py:
```diff
@@ -0,0 +1,117 @@
+        if identified_release is not None:
+            return True, identified_release
+
+        return False
```
Return `False, None` to have a unified return value pattern.

ohadravid commented on src/headphones2/postprocess/taggers/acoustid_tagger.py:
```diff
@@ -0,0 +1,117 @@
+    def _acoustid_tag_file(filepath):
+        """
+        Gets metadata for a file from Acoustid.
+        returns a Result object with Fingerprint, acoustid, recording_ids and release_ids
```
Should document that any exception is suppressed and None items may be returned.

ohadravid commented on src/headphones2/postprocess/taggers/simple_tagger.py:
```diff
@@ -0,0 +1,28 @@
+logger = logbook.Logger(__name__)
+
+
+class SimpleBeetsTagger(PostProcessorComponentBase):
```
Why is it `Simple`? Is there an advanced version?

ohadravid commented on src/headphones2/postprocess/component_base.py:
```diff
@@ -0,0 +1,13 @@
+import abc
+
+
+class PostProcessorComponentBase(object):
```
`ComponentBase` is redundant since it is ABCMeta and any class is a componenet...

ohadravid commented on test/test_acoustid_tagger.py:
```diff
@@ -0,0 +1,38 @@
+    for item in items:
+        item.acoustid_id = ''
+        item.acoustid_fingerprint = ''
+        item.write()
```
If you change the code so that after the `process` we do a `write` on each item.
It will move the IO to the top of the code, it will be easier to test (Create an item without the ids, check after the process that they are added. never write them to disk).
See http://www.youtube.com/watch?v=DJtef410XaM.

omerbenamram replied:
will implement this in PostProcessor base class

ohadravid commented on test/test_acoustid_tagger.py:
```diff
@@ -0,0 +1,38 @@
+
+@pytest.fixture
+def single_item():
+    track_22 = r"C:\Users\Omer\Desktop\The Theory of Everything [2013]\20 Ayreon - The Theory of Everything [2013] - Phase 2_ Symmetry, Dark Medicine.flac"
```
Maybe some open-source music file which is properly detected by both them, and add it to the git repo so everyone can run the tests.

omerbenamram committed:
changed some of ohad's comments

omerbenamram committed:
__future__ everywhere!!!

omerbenamram committed:
more changes

omerbenamram committed:
completed work on renamer, added feature to force a single output folder

omerbenamram committed:
process and write are now separate methods for renamer

omerbenamram committed:
finished going over ohad's comments.
tests are still broken

omerbenamram committed:
small change

omerbenamram committed:
actually using separate write methods now :)

omerbenamram committed:
acoustid is now invoked from inside beets tagger

omerbenamram commented:
i've gotten this to a nice working state.
take a look

however, tests are still very lacking :(

omerbenamram committed:
added documentation

ohadravid commented on src/headphones2/postprocess/process_folder.py:
```diff
@@ -0,0 +1,72 @@
+    beets_tagger = BeetsTagger()
+    renamer = Renamer()
+
+    for tagger in [aid_tagger, beets_tagger, renamer]:
```
Why not just `post_processors = [aid_tagger, beets_tagger, renamer]`?

ohadravid commented on src/headphones2/postprocess/process_folder.py:
```diff
@@ -0,0 +1,72 @@
+    logger.info("Post processor compelted for {}".format(folder))
+
+
+def _write_all_changes(list_of_items, list_of_post_processors=(), **kwargs):
```
Why default value?

ohadravid commented on src/headphones2/postprocess/taggers/simple_tagger.py:
```diff
@@ -0,0 +1,65 @@
+
+        return BeetsTaggerResult(album_info.album_id, album_info, track_mapping, extra_items, extra_tracks)
+
+    def write(self, item_list, album_info=None, track_mapping=None, **kwargs):
```
Can you use the internal state of the class to do that?
Why have a write at all? Every call re-implement this bit of code every time.
The kwargs handling is very strange, and independent classes might get the wrong data, confuse arguments etc..
Also, how do you know which kwargs to pass?

omerbenamram committed:
changed taggers to use a common task object.

omerbenamram committed:
moved task to a separate file. fixed a small bug

omerbenamram commented:
task state is now saved in an AlbumTask object!
this means no write methods, each processor applies pure logic to the task.

i'm still wondering about how extensible this class should be, and what API's it should expose.

what do you think?

omerbenamram committed:
very basic tests for post processor!
uploaded sample files from http://freemusicarchive.org/
metadata is now written before renamer is called (bugfix)
we can now pass destination folder properly

omerbenamram committed:
small fix

omerbenamram commented:
FINALLY some working tests

omerbenamram committed:
fix

omerbenamram committed:
2 small fixed.
1 - proper exception text now thrown
2 - py3 compat

omerbenamram committed:
pre-processor now copies files
